# Exploratory Data and Feature Analysis FCDO

This notebook is meant to roughly explore the features and their relation with conflict outbreak. We try to infer which pre-preocessing steps are necessary in order to fit a random forest model. Also we use correlation coefficients to assess their ability to explain conflict.

In [66]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import logging.config
import difflib as dl

In [62]:
pd.set_option('display.max_columns', 500)
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format

In [36]:
# Import data
root = './Data/'
file = 'FCDO_data.csv'
data = pd.read_csv(root+file)

## Exploratory Analysis

In [37]:
data.sample(5)

ADM3_EN  IDAHO_EPSCOR_TERRACLIMATE_Max_Climate_Water_Deficit  \
189           Markaz Ana                                               3136     
162  Markaz Al-Kadhimiya                                               3064     
221       Markaz Tilkaef                                               2856     
60           Al-Muotasim                                               3238     
76            Al-Saadiya                                               3335     

     IDAHO_EPSCOR_TERRACLIMATE_SD_Max_Climate_Water_Deficit  \
189                                          71.175878        
162                                          49.875841        
221                                          41.614693        
60                                          123.040995        
76                                          154.774969        

     IDAHO_EPSCOR_TERRACLIMATE_Mean_Climate_Water_Deficit  \
189                                        1588.232887      
162                                        1560.738971      
221                                        1057.869235      
60                                         1547.291400      
76                                         1490.178490      

    IDAHO_EPSCOR_TERRACLIMATE_SD_Mean_Climate_Water_Deficit  \
189                                 42.992616920492175        
162                                  11.85578428443833        
221                                  38.96659814083428        
60                                   27.80471970536975        
76                                   42.72876919029818        

     IDAHO_EPSCOR_TERRACLIMATE_Min_Climate_Water_Deficit  \
189                                                169     
162                                                203     
221                                                  0     
60                                                 155     
76                                                  37     

     IDAHO_EPSCOR_TERRACLIMATE_SD_Min_Climate_Water_Deficit  \
189                                          66.902788        
162                                           7.116356        
221                                           0.000000        
60                                           44.018793        
76                                           56.888399        

     ECMWF_ERA5_LAND_MONTHLY_Max_Evaporation_Bare  \
189                                  0.000000e+00   
162                                 -1.489324e-05   
221                                 -2.089748e-04   
60                                  -1.047505e-05   
76                                  -3.003515e-07   

    ECMWF_ERA5_LAND_MONTHLY_SD_Max_Evaporation_Bare  \
189                           4.137043297674831e-07   
162                            6.73731196573838e-07   
221                           3.542985503604047e-05   
60                           1.0347014324383653e-05   
76                           2.0563184139741513e-05   

     ECMWF_ERA5_LAND_MONTHLY_Max_Evaporation_Potential  ...  \
189                                          -0.001448  ...   
162                                          -0.002253  ...   
221                                          -0.001792  ...   
60                                           -0.001601  ...   
76                                           -0.001011  ...   

    Strategic developments  Violence against civilians total_event_types  \
189                     11                           0                19   
162                      3                           9                37   
221                      7                           3                24   
60                       1                           1                 8   
76                       4                           4                31   

     ('fatalities', 'Battles') ('fatalities', 'Explosions/Remote violence')  \
189                          0                                            8   
162                  

In [38]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 268 entries, 0 to 267
Columns: 389 entries, ADM3_EN to ('total_fatalities', '')
dtypes: float64(203), int64(50), object(136)
memory usage: 814.6+ KB


There are 136 object types but it turns out they are not objects at all but just stored as objects because they contain '--', so we replace that by nan values. After that we convert everything (except administrative district) to float:

In [39]:
# replace -- by np.nan
data.replace(['--'], [np.nan], inplace=True)  # replace -- by np.nan
data.loc[:, data.columns != 'ADM3_EN'] = data.loc[:, data.columns != 'ADM3_EN'].apply(pd.to_numeric, errors='raise')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 268 entries, 0 to 267
Columns: 389 entries, ADM3_EN to ('total_fatalities', '')
dtypes: float64(338), int64(50), object(1)
memory usage: 814.6+ KB


In [40]:
data.describe()

IDAHO_EPSCOR_TERRACLIMATE_Max_Climate_Water_Deficit  \
count                                         268.000000     
mean                                         3078.511194     
std                                           278.466702     
min                                          2145.000000     
25%                                          2974.000000     
50%                                          3184.500000     
75%                                          3250.250000     
max                                          3757.000000     

       IDAHO_EPSCOR_TERRACLIMATE_SD_Max_Climate_Water_Deficit  \
count                                         268.000000        
mean                                           99.042011        
std                                            96.542933        
min                                            20.233786        
25%                                            57.420354        
50%                                            85.088010        
75%                                           120.806300        
max                                          1460.724776        

       IDAHO_EPSCOR_TERRACLIMATE_Mean_Climate_Water_Deficit  \
count                                         267.000000      
mean                                         1420.084923      
std                                           342.119412      
min                                           629.419212      
25%                                          1174.810897      
50%                                          1543.433560      
75%                                          1671.400007      
max                                          2016.981785      

       IDAHO_EPSCOR_TERRACLIMATE_SD_Mean_Climate_Water_Deficit  \
count                                         267.000000         
mean                                           44.954994         
std                                            21.483115         
min                                             7.863495         
25%                                            29.344222         
50%                                            41.205180         
75%                                            55.720365         
max                                           130.977805         

       IDAHO_EPSCOR_TERRACLIMATE_Min_Climate_Water_Deficit  \
count                                         268.000000     
mean                                          117.865672     
std                                           154.351454     
min                                             0.000000     
25%                                             0.000000     
50%                                             0.000000     
75%                                           212.000000     
max                                           536.000000     

       IDAHO_EPSCOR_TERRACLIMATE_SD_Min_Climate_Water_Deficit  \
count                                         268.000000        
mean                                           29.129031        
std                                            29.852595        
min                                             0.000000        
25%                                             0.000000        
50%                                            22.779419        
75%                                            45.751868        
max                                           167.854296        

       ECMWF_ERA5_LAND_MONTHLY_Max_Evaporation_Bare  \
count                                  2.650000e+02   
mean                                  -3.762279e-05   
std                                    6.841757e-05   
min                                   -3.488157e-04   
25%                                   -3.040070e-05   
50%                                   -8.313451e-06   
75%                                   -9.872019e-08   
max                                    0.000000e+00   

       ECMWF_ERA5_LAND_MONTHLY_SD_Max_Evaporation_Bare  \


In [41]:
#Strip ),' and " from column names cause otherwise column selection is annoying
data.columns = data.columns.str.replace("'","")
data.columns = data.columns.str.replace("''","")
data.columns = data.columns.str.replace("(","")
data.columns = data.columns.str.replace(")","")

/tmp/ipykernel_49/2893551656.py:4: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data.columns = data.columns.str.replace("(","")
/tmp/ipykernel_49/2893551656.py:5: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data.columns = data.columns.str.replace(")","")


## Including Ranges

In [42]:
# Find columns minimums and maximums. (Note: there are no min/max columns (without capital letter))
min_cols = [col for col in data.columns if 'Min' in col]
max_cols = [col for col in data.columns if 'Max' in col]
print(len(max_cols), len(min_cols))

122 122


In [43]:
import difflib as dl
# Find which column from min corresponds with max: #https://docs.python.org/3/library/difflib.html#difflib.get_close_matches
for min_col in min_cols:
    max_col = dl.get_close_matches(min_col, max_cols, n=1)[0]
    rangename=min_col.replace('Min','Range')
    data[rangename] = data[max_col]-data[min_col]

/tmp/ipykernel_49/1446068032.py:6: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  data[rangename] = data[max_col]-data[min_col]


In [44]:
data.head(5)

ADM3_EN  IDAHO_EPSCOR_TERRACLIMATE_Max_Climate_Water_Deficit  \
0  Abi Gharaq                                               3246     
1    Abu Dalf                                               3254     
2  Abu Ghraib                                               3206     
3    Agjalare                                               2771     
4      Akaika                                               3239     

   IDAHO_EPSCOR_TERRACLIMATE_SD_Max_Climate_Water_Deficit  \
0                                          72.455552        
1                                          66.514535        
2                                          96.117233        
3                                          72.290144        
4                                          40.780722        

   IDAHO_EPSCOR_TERRACLIMATE_Mean_Climate_Water_Deficit  \
0                                        1698.439815      
1                                        1573.475734      
2                                        1602.843255      
3                                        1049.910216      
4                                        1815.596549      

   IDAHO_EPSCOR_TERRACLIMATE_SD_Mean_Climate_Water_Deficit  \
0                                          24.124774         
1                                          23.370776         
2                                          28.625457         
3                                          43.778659         
4                                          32.822756         

   IDAHO_EPSCOR_TERRACLIMATE_Min_Climate_Water_Deficit  \
0                                                408     
1                                                 27     
2                                                203     
3                                                  0     
4                                                391     

   IDAHO_EPSCOR_TERRACLIMATE_SD_Min_Climate_Water_Deficit  \
0                                          22.849043        
1                                          45.195230        
2                                          44.268045        
3                                           0.000000        
4                                          30.965841        

   ECMWF_ERA5_LAND_MONTHLY_Max_Evaporation_Bare  \
0                                     -0.000004   
1                                     -0.000010   
2                                     -0.000015   
3                                     -0.000263   
4                                     -0.000004   

   ECMWF_ERA5_LAND_MONTHLY_SD_Max_Evaporation_Bare  \
0                                         0.000003   
1                                         0.000007   
2                                         0.000011   
3                                         0.000062   
4                                         0.000003   

   ECMWF_ERA5_LAND_MONTHLY_Max_Evaporation_Potential  ...  \
0                                          -0.003034  ...   
1                                          -0.001444  ...   
2                                          -0.002253  ...   
3                                          -0.001468  ...   
4                                          -0.002439  ...   

   NASA_NOAH01_Range_Surface_Wind_Speed  \
0                              3.390722   
1                              3.525623   
2                              3.605780   
3                              1.630532   
4                              3.699778   

   NASA_NOAH01_SD_Range_Surface_Wind_Speed  \
0                                -0.037728   
1                                 0.120121   
2                                 0.038813   
3                                -0.029991   
4                                 0.080373   

   inuncoast_historical_nosub_hist_Coastal_Flood_Range  \
0                                                  0     
1                                                  0     
2                                                  0     
3      

## Correlation Analysis:

In [74]:
non_prediction_variables = ['ADM3_EN','total_fatalities, ']

In [75]:
X = data.loc[:,~(data.columns.isin(non_prediction_variables))]
X.columns

Index(['IDAHO_EPSCOR_TERRACLIMATE_Max_Climate_Water_Deficit',
       'IDAHO_EPSCOR_TERRACLIMATE_SD_Max_Climate_Water_Deficit',
       'IDAHO_EPSCOR_TERRACLIMATE_Mean_Climate_Water_Deficit',
       'IDAHO_EPSCOR_TERRACLIMATE_SD_Mean_Climate_Water_Deficit',
       'IDAHO_EPSCOR_TERRACLIMATE_Min_Climate_Water_Deficit',
       'IDAHO_EPSCOR_TERRACLIMATE_SD_Min_Climate_Water_Deficit',
       'ECMWF_ERA5_LAND_MONTHLY_Max_Evaporation_Bare',
       'ECMWF_ERA5_LAND_MONTHLY_SD_Max_Evaporation_Bare',
       'ECMWF_ERA5_LAND_MONTHLY_Max_Evaporation_Potential',
       'ECMWF_ERA5_LAND_MONTHLY_SD_Max_Evaporation_Potential',
       ...
       'NASA_NOAH01_Range_Surface_Wind_Speed',
       'NASA_NOAH01_SD_Range_Surface_Wind_Speed',
       'inuncoast_historical_nosub_hist_Coastal_Flood_Range',
       'inuncoast_historical_nosub_hist_Coastal_Flood_SD_Range',
       'inunriverine_historical_nosub_hist_Riverine_Flood_Range',
       'inunriverine_historical_nosub_hist_Coastal_Riverine_SD_Range',
       'h

In [56]:
# Make correlation plot to find out what variables are most correlated to total_fatalities
correlations = data.corr()

In [64]:
# Take abs value since both negative and positive correlations have predictive/causal power to have an idea which variables are most related.
correlations['abs_total_fatalities']= np.abs(correlations['total_fatalities, '])
correlations.sort_values('abs_total_fatalities', ascending=False)['abs_total_fatalities']

total_fatalities,                                                      1.00
fatalities, Battles                                                    0.94
fatalities, Explosions/Remote violence                                 0.90
Battles                                                                0.90
total_event_types                                                      0.70
Explosions/Remote violence                                             0.68
Strategic developments                                                 0.62
IDAHO_EPSCOR_TERRACLIMATE_SD_Min_Precipitation_Accumulation            0.60
ECMWF_ERA5_SD_LAND_MONTHLY_Mean_Skin_Reservoir                         0.47
ECMWF_ERA5_SD_LAND_MONTHLY_Min_Skin_Reservoir                          0.47
fatalities, Violence against civilians                                 0.46
Violence against civilians                                             0.43
ECMWF_ERA5_LAND_MONTHLY_Mean_Skin_Reservoir                            0.41
NASA_NOAH01_